src/edgepi/calibration/edgepi_eeprom.py

```diff
@@ -171,3 +181,128 @@ def page_write_register(self, mem_addr: int = None, data: list = None):
+    def __parameter_sanity_check(self, mem_addr: int = None,
+                                 length: int = None,
+                                 user_space: bool = True):
+        """
+        Generic check function for target memory address and length of data to read/write
+        Args:
+            mem_addr (int): starting memory address to read from
+            length (int): length of data to read
+            user_space (bool): True if mem_addr is in user space
+        """
+        # End address depending on the user_space flag
+        end_address = EdgePiMemoryInfo.USER_SPACE_END_BYTE.value if user_space \
+                      else (EdgePiMemoryInfo.USER_SPACE_START_BYTE.value - 1)
+        # Checks whether proper values are passed
+        if mem_addr is None or length is None or mem_addr < 0 or length <= 0:
+            raise ValueError(f'Invalid Value passed: {mem_addr}, {length}')
+        # Checks whether starting address and length of data to read/write are within memory bound
+        if mem_addr+length > end_address:
+            raise MemoryOutOfBound(f'Operation range is over the size of the memory by \
+                                     {mem_addr+length-end_address}')
+
+    def __generate_list_of_pages(self, mem_addr: int = None, data: list = None):
+        """
+        Generate a two dimensional structured list with length of a page. This is method is used for
+        read/write by page
+        Args:
+            mem_addr (int): starting memory address to read from
+            data (list): data to write
+        Return:
+            page_writable_list (list): [[Page_N], [Page_N+1]...]]
+        """
+        # starting memory address can be anywhere from 0~63, check whether the length of data fits
+        # within the range.
+        if (mem_addr%EEPROMInfo.PAGE_SIZE.value + len(data)) < EEPROMInfo.PAGE_SIZE.value:
+            page_n = [data]
+        else:
+            curr_page_remainder = EEPROMInfo.PAGE_SIZE.value - mem_addr%EEPROMInfo.PAGE_SIZE.value
+            page_1 = [data[val] for val in range(curr_page_remainder)]
+            data_remainder = data[curr_page_remainder:]
+            page_n = [data_remainder[byte:byte+EEPROMInfo.PAGE_SIZE.value] \
+                      for byte in range(0,len(data_remainder), EEPROMInfo.PAGE_SIZE.value)]
+            page_n.insert(0, page_1)
+        return page_n
+
+    def read_memory(self, start_addrx: int = None, length: int = None):
+        """
+        Read user space memory starting from 0 to 16383
+        Args:
+            mem_addr (int): starting memory address to read from
+            length (int): length of data to read
+        Return:
+            data (list): list of data read from the specified memory and length
+        """
+        self.__parameter_sanity_check(start_addrx, length, True)
+        dummy_data = self.__generate_list_of_pages(start_addrx, [0]*length)
+        start_addrx = start_addrx + EdgePiMemoryInfo.USER_SPACE_START_BYTE.value
```

Comment:
why is there an x at the end of `start_addrx`

Reply:
didn't want to spell out the address but wanted it to sound similar 😅